### ISSUE TITLE
Knex Query Wrapper

### ISSUE THREAD

**Issue by bgphipps** [tgriesser/knex#1278]
Hey all,

I need to decamelize my columns on all my queries and was thinking it would be best to do this at the lowest level possible. Does anyone have any good suggestions for doing this as a wrapper to knex?

Example:
`insert into `table` (`colOne`, `colTwo`)  values (1, 2)`

would become 

`insert into `table` (`col_one`, `col_two`)  values (1, 2)`

It would also be useful to camelize on the way out as well.

Thanks!

**Comment by sheeley** [comment]
It looks like, based on this line:

https://github.com/tgriesser/knex/blob/7cc827f6cf495faa128b0b881d2ec4d6aa454fd9/src/query/compiler.js#L20

We may be able to pass in a `formatter` here?

**Comment by SamDuvall** [comment]
@bgphipps I too have gone done this route and would love a solution like what you're talking about. I have sort of accomplished this by decorating the query object and intercepting the various call (e.g. where, update, insert, etc). However, that solution stinks, because it's fragile and it requires a lot of repetition. @sheeley I like the idea of doing something with the formatter. I might play around with that and see if I come up with anything.

**Comment by wubzz** [comment]
I personally don't see how this belongs in knex's code base, but I don't have a strong opinion against it either.

For adding something like this I would recommend adding it to the formatter's [columnize](https://github.com/tgriesser/knex/blob/master/src/formatter.js#L15) function as that should be the best place in the code base.

I'm just springing ideas now but to make it as dynamic as possible, maybe something like `.transformColumns()` could be nice.

```javascript
knex('table').insert(data).transformColumns((column) => { return decamelize(column); });
```
_(For lack of a better name)_

**Comment by SamDuvall** [comment]
@wubzz I agree. I'm thinking of something that wraps knex.

**Comment by sheeley** [comment]
@wubzz totally agree - don't need transformation to be part of knex, but hooks to allow it would be sweet.

i'm trying to abstract things between our frontend & api devs (like @bgphipps) and our DBAs, so if there were a way to do something like this and have hooks into `where`, `join`, etc that'd be awesome. Right now, we're manually wrapping everything, and it works but isn't a joy.

```js
const readableToColumn = {
  objectName: `obj_nm`,
  objectId: `obj_id`
}

const columnToReadable = someReverse(columnNames)

knex.registerTransformIn((row) => {
  const out = {}
  for (const key of Object.keys(row)) {
    out[readableToColumn[key] || key] = row[key]
  }
})

knex.registerTransformOut((row) => {
  const out = {}
  for (const key of Object.keys(row)) {
    out[columnToReadable[key] || key] = row[key]
  }
})

const objectId = [1, 2, 3]
knex.select([`objectName`, `objectId`]).where({objectId})
// => [{objectName: `object 1`, objectId: 1}, ...]
```

**Comment by kripod** [comment]
I have created a "debug helper" for knex/lib/formatter.js, which prints out console messages about the return value of each function: https://gist.github.com/kripod/127d03e598d3840e57d9755d65283ab3

Basically, it seems that the `wrap` and `_wrapString` functions are responsible for processing fields which should be queried. By the way, I am [currently working on a Knex query wrapper](https://github.com/kripod/knex-orm/blob/master/src/query-builder.js) for my soon-to-be-released project called knex-orm.

**Comment by kripod** [comment]
I have just completed my little query wrapper for my ORM today, see the following code parts for my solution:

https://github.com/kripod/knex-orm/blob/f852fcb09b041ffeb79a1df16eb8fd76e5bc34ca/src/index.js#L32

https://github.com/kripod/knex-orm/blob/f852fcb09b041ffeb79a1df16eb8fd76e5bc34ca/src/query-builder.js#L42

**Comment by sheeley** [comment]
Thanks, that looks super useful!

**Comment by sheeley** [comment]
Translating INTO the db: 
```js
export const db = knex(config)

// Override knex formatter to translate columns from appId to app_id, etc.
const formatterPrototype = db.client.Formatter.prototype
const originalWrapString = formatterPrototype._wrapString
formatterPrototype._wrapString = function _wrapString(value) {
  return originalWrapString.call(this, readableToColumn[value] || value)
}
```

**Comment by zquancai** [comment]
In addition, I need to camelize queries results. I do like this:
```js
const originQuery = .knex.client.query;
  knex.client.query = function query(connection, obj) {
    const promise = originQuery.call(this, connection, obj);
    promise.then(function(obj) {
      obj.response[0] =  camelizeKeys(obj.response[0]);
      return obj;
    });
    return promise;
  }
```
It run normally, do you think it's a good solution?

**Comment by razvanz** [comment]
@wubzz @SamDuvall  I don't think defining a specific function for translating columns is a good idea.

A suggested alternative is having a `format` / `parse` interface at the QueryBuilder level (or even at the root level: `makeKnex`), where a user could register a function for modifying the query (`format`) and the result (`parse`):

```
let db = (table) => knex('table')
  .format((q) => { q.sql = snakeify(q.sql); return q })
  .parse((res) => camelify(res))

db('testTable')
  .column('colOne', 'colTwo')
  .select()
  .then(console.log)
```

This would be a more generic approach which would allow other kinds of transformations while keeping the interface composable.

**Comment by elhigu** [comment]
@razvanz This cannot be done: `snakeify(q.sql)` you would have to implement pretty much complete SQL parser to know what parts of string must be transformed. 

So transformation must be done at some point when we still know that certain string is identifier etc. so hook would be probably something like `formatIdentifier(colName => snakeify(colName))`.

**Comment by razvanz** [comment]
@elhigu How about the parsing part? I came to this issue interested in having `parse` hook, commonly registered for `then` / `stream` / `asCallback` / etc.

**Comment by elhigu** [comment]
I would be surprised if there isn't already an opened issue for that feature, but sure create a new one if it doesn't exist yet.

**Comment by lehni** [comment]
It looks like the new `wrapIdentifier ()` config option will be perfect for this:

https://github.com/tgriesser/knex/blob/9658c78d4fba55a129acc3843b5150de93d553c4/src/client.js#L175

```js
const knex = Knex({
  wrapIdentifier(value, wrapIdentifier) {
    return wrapIdentifier(underscore(value))
  }
})
```

This will only be available once 0.14.0 is out...
